packages/nextjs/src/config/loaders/ast.ts

```diff
@@ -320,3 +328,169 @@ export function removeComments(ast: AST): void {
   const nodesWithComments = ast.find(Node).filter(nodePath => !!nodePath.node.comments);
   nodesWithComments.forEach(nodePath => (nodePath.node.comments = null));
 }
+
+/**
+ * Determines from a given AST of a file, whether the file has a default export or not.
+ */
+export function hasDefaultExport(ast: AST): boolean {
+  const hasDefaultDeclaration = ast.find(ExportDefaultDeclaration).size() > 0;
```

Comment:
Pretty sure we don't need a function for this.

```suggestion
  const hasDefaultDeclaration = ast.find(ExportDefaultDeclaration).length > 0;
```

Also, I wonder if we could simplify this function by doing something like this:

```js
export function hasDefaultExport(ast: AST): boolean {
  const defaultExports = ast.filter(nodePath => 
    ExportDefaultDeclaration.check(nodePath.node) 
    || ExportDefaultSpecifier.check(nodePath.node)
    || (ExportSpecifier.check(nodePath.node) && nodePath.node.exported.name === 'default')
   );

  // In theory there should only ever be 0 or 1, but who knows what people do
  return defaultExports.length > 0;
}
```

WDYT?

Reply:
As for the size vs. length: sure, shouldn't make a difference (looking at jscodeshift code, they do exactly the same thing lol)

```js
/**
   * Returns the number of elements in this collection.
   *
   * @return {number}
   */
  size() {
    return this.__paths.length;
  }

  /**
   * Returns the number of elements in this collection.
   *
   * @return {number}
   */
  get length() {
    return this.__paths.length;
  }
```

As for the snippet, it doesn't work since `filter` doesn't recurse the node tree. It just looks at child paths.

The following would work, but I think that this is quadratic complexity and that would be catastrophic in larger files. I am actually not a 100% sure on that one since the jscodeshift code around that is a tiny bit complex but I'd rather not risk it:

```ts
export function hasDefaultExport(ast: AST): boolean {
  const defaultExports = ast.find(Node, value => {
    return (
      ExportDefaultDeclaration.check(value) ||
      ExportDefaultSpecifier.check(value) ||
      (ExportSpecifier.check(value) && value.exported.name === 'default')
    );
  });

  // In theory there should only ever be 0 or 1, but who knows what people do
  return defaultExports.length > 0;
}
```

I think I'm gonna stick with what we have because we know it doesn't have bad complexity and you can understand what it does at first sight.

Reply:
> As for the snippet, it doesn't work since filter doesn't recurse the node tree. It just looks at child paths.

Oh, damn, you're right - I was thinking of it operating on the results of a `find` call (which is just an array of `nodePath`s), not on the full AST. Good catch.